File: py/codesamples/model/model_client.py
"""
Copyright (c) Aalyria Technologies, Inc., and its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import argparse
import sys
from io import BytesIO
from pathlib import Path

import api.model.v1.model_pb2 as model_pb2
import api.model.v1.model_pb2_grpc as model_pb2_grpc
import grpc
import nmts.v1.proto.nmts_pb2 as nmts_pb2

from py.authentication import auth


def list_entities(stub: model_pb2_grpc.ModelStub, page_size: int = 0) -> list[nmts_pb2.Entity]:
  """Reads every entity in the model, one page at a time.

  ListEntities answers with a page of entities and, when more remain, a token
  naming where the next page begins. A caller that wants the whole model has to
  follow those tokens until the response carries none, which is what the loop
  below does. Passing a page_size of 0 lets the server choose the page size.
  """
  entities = []
  request = model_pb2.ListEntitiesRequest(page_size=page_size)
  while True:
    response = stub.ListEntities(request)
    entities.extend(response.entities)
    if not response.next_page_token:
      return entities
    # A cursor that fails to advance would walk the same page forever. Returning
    # the entities read so far would look like the whole model to a caller that
    # acts on it, so this fails instead.
    if response.next_page_token == request.page_token:
      raise RuntimeError(
        f"ListEntities repeated the page token it was given ({len(entities)} entities read); the walk cannot advance"
      )
    request.page_token = response.next_page_token


def establish_connection(target: str, email: str, key_id: str, private_key: str) -> model_pb2_grpc.ModelStub:
  # Create auth config
  config = auth.Config(email=email, private_key_id=key_id, private_key=BytesIO(private_key.encode("utf-8")))

  # Create call credentials
  call_credentials = auth.new_credentials(config)

  # Combine with SSL channel credentials
  channel = grpc.secure_channel(
    target,
    grpc.composite_channel_credentials(grpc.ssl_channel_credentials(), call_credentials),
    [
      (
        "grpc.max_receive_message_length",
        1024 * 1024 * 256,
      ),
    ],
  )
  return model_pb2_grpc.ModelStub(channel)


def main():
  # Setup argparser
  parser = argparse.ArgumentParser()
  parser.add_argument(
    "target",
    type=str,
    help="The target URL of the Spacetime Model API (e.g., 'api.example.com' or 'api.example.com:8080').",
  )
  parser.add_argument(
    "email",
    type=str,
    help="Client Email for Spacetime Auth.",
  )
  parser.add_argument(
    "key_id",
    type=str,
    help="Client Key ID for Spacetime Auth.",
  )
  parser.add_argument(
    "private_key_path",
    type=str,
    help="The Client Key File Path for Spacetime Auth.",
  )
  parser.add_argument(
    "--page_size",
    type=int,
    default=0,
    help="The number of entities to request per page. 0 lets the server choose.",
  )
  args = parser.parse_args()

  # The private key should start with "-----BEGIN RSA PRIVATE KEY-----" and
  # end with "-----END RSA PRIVATE KEY-----". In between, there should be newline-delimited
  # strings of characters.
  private_key = Path(args.private_key_path).read_text()

  stub = establish_connection(args.target, args.email, args.key_id, private_key)
  entities = list_entities(stub, args.page_size)
  print("ListEntitiesResponse received:\n", entities)


if __name__ == "__main__":
  main()

File: go/codesamples/model/model_client.go
// Copyright (c) Aalyria Technologies, Inc., and its affiliates.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"math"
	"os"
	"strings"
	"time"

	"github.com/jonboulle/clockwork"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/keepalive"

	nmts "outernetcouncil.org/nmts/v1/proto"

	model "aalyria.com/spacetime/api/model/v1"
	"aalyria.com/spacetime/auth"
)

const (
	maxMessageSize    = 256 * 1024 * 1024
	keepAliveTime     = 30 * time.Second
	keepAliveTimeout  = 10 * time.Second
	keepAliveInterval = 5 * time.Second
)

// NewCredentialsFromPrivateKey creates gRPC credentials from a private key string.
func NewCredentialsFromPrivateKey(email, privateKeyID, privateKey string) (credentials.PerRPCCredentials, error) {
	// Create auth config using the auth package
	authConfig := auth.Config{
		Clock:        clockwork.NewRealClock(),
		PrivateKey:   strings.NewReader(privateKey),
		PrivateKeyID: privateKeyID,
		Email:        email,
	}

	// Create credentials using the auth package
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	return auth.NewCredentials(ctx, authConfig)
}

// listPageTimeout bounds one page of a walk. A whole model takes as many
// requests as it has pages, so a deadline covering the walk would have to grow
// with the model; one that covers a single page does not.
const listPageTimeout = 30 * time.Second

// listEntities calls the Model API to list entities, walking every page and returning the entities from all of them.
// The page size comes from the -page_size flag; 0 lets the server choose.
func listEntities(ctx context.Context, client model.ModelClient, pageSize int32) ([]*nmts.Entity, error) {
	var entities []*nmts.Entity

	req := &model.ListEntitiesRequest{PageSize: pageSize}

	for {
		pageCtx, cancel := context.WithTimeout(ctx, listPageTimeout)
		resp, err := client.ListEntities(pageCtx, req)
		cancel()
		if err != nil {
			return nil, fmt.Errorf("failed to list entities: %w", err)
		}
		entities = append(entities, resp.GetEntities()...)
		if resp.GetNextPageToken() == "" {
			return entities, nil
		}
		// A cursor that fails to advance would walk the same page forever.
		// Returning the entities read so far would look like the whole model
		// to a caller that acts on it, so this fails instead.
		if resp.GetNextPageToken() == req.PageToken {
			return nil, fmt.Errorf("ListEntities repeated the page token it was given (%d entities read); the walk cannot advance", len(entities))
		}
		req.PageToken = resp.GetNextPageToken()
	}
}

// establishConnection creates a gRPC connection to the Model API with authentication.
func establishConnection(target, email, keyID, privateKey string) (*grpc.ClientConn, error) {
	// Create call credentials
	callCreds, err := NewCredentialsFromPrivateKey(email, keyID, privateKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create call credentials: %w", err)
	}

	// Create composite credentials with TLS
	creds := credentials.NewTLS(nil)
	compositeCreds := grpc.WithPerRPCCredentials(callCreds)

	// Build connection options
	opts := []grpc.DialOption{
		grpc.WithTransportCredentials(creds),
		compositeCreds,
		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMessageSize)),
		grpc.WithKeepaliveParams(keepalive.ClientParameters{
			Time:                keepAliveTime,
			Timeout:             keepAliveTimeout,
			PermitWithoutStream: true,
		}),
	}

	// Connect to the server
	conn, err := grpc.NewClient(target, opts...)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to %s: %w", target, err)
	}

	return conn, nil
}

func run() error {
	// Parse command line arguments
	var (
		target         = flag.String("target", "", "The target URL of the Spacetime Model API (e.g., 'api.example.com' or 'api.example.com:8080')")
		email          = flag.String("email", "", "Client email for Spacetime authentication")
		keyID          = flag.String("key_id", "", "Client key ID for Spacetime authentication")
		privateKeyPath = flag.String("private_key_path", "", "Path to the private key file")
		pageSize       = flag.Int("page_size", 1000, "Page size to limit query to; default = 1000")
	)
	flag.Parse()

	// Validate required arguments
	if *target == "" || *email == "" || *keyID == "" || *privateKeyPath == "" {
		fmt.Fprintf(os.Stderr, "Usage: %s -target <target> -email <email> -key_id <key_id> -private_key_path <private_key_path>\n", os.Args[0])
		flag.PrintDefaults()
		return fmt.Errorf("missing required arguments")
	}

	// Validate page size argument
	if *pageSize < 0 || *pageSize > math.MaxInt32 {
		return fmt.Errorf("page size is expected to be between 0 and %v, received: %v", math.MaxInt32, *pageSize)
	}

	// Read private key
	privateKeyBytes, err := os.ReadFile(*privateKeyPath)
	if err != nil {
		return fmt.Errorf("error reading private key file: %w", err)
	}
	privateKey := string(privateKeyBytes)

	// Establish connection
	conn, err := establishConnection(*target, *email, *keyID, privateKey)
	if err != nil {
		return fmt.Errorf("error establishing connection: %w", err)
	}
	defer conn.Close()

	// Create client
	client := model.NewModelClient(conn)

	// Each page of the walk below is bounded by its own deadline, so the walk
	// itself runs for as long as the model needs.
	ctx := context.Background()

	// List entities
	entities, err := listEntities(ctx, client, int32(*pageSize))
	if err != nil {
		return fmt.Errorf("error listing entities: %w", err)
	}

	// Print results
	fmt.Printf("ListEntitiesResponse received:\n%v", entities)
	return nil
}

func main() {
	if err := run(); err != nil {
		log.Fatalf("%v", err)
	}
}
